README.md
# step_slider

Flutter slider widget that snaps to given step values and notifies about step changes.

![StepSlider demo](https://giant.gfycat.com/LongFailingDowitcher.gif)

## Getting Started

Add StepSlider to your widget tree along with needed custom arguments:

```dart
@override
Widget build(BuildContext context) {
    ...
    child: StepSlider(
        min: 100.0,
        max: 200.0,
        steps: {110, 150, 160, 195},
        initialStep: 150,
        animCurve: Curves.bounceInOut,
        animDuration: Duration(seconds: 1),
        snapMode: SnapMode.value(10),
        hardSnap: true,
        onStepChanged: (it) => print('Step changed to $it.'),
        // ... slider's other args
    ),
}
```


CHANGELOG.md
## [1.0.0+1] - 15/04/2019

* Implement core functionality.


lib/src/slider.dart
import 'package:flutter/material.dart';
import 'package:step_slider/src/snap_mode.dart';
import 'package:step_slider/src/snapper.dart';

class StepSlider extends StatefulWidget {
  StepSlider({
    Key key,
    this.sliderKey,
    @required this.steps,
    this.initialStep,
    this.onStepChanged,
    this.snapMode,
    this.hardSnap: false,
    this.animCurve: Curves.fastLinearToSlowEaseIn,
    this.animDuration: const Duration(milliseconds: 500),
    this.min: 0.0,
    this.max: 1.0,
    this.label,
    this.divisions,
    this.onChanged,
    this.onChangeEnd,
    this.onChangeStart,
    this.activeColor,
    this.inactiveColor,
    this.semanticFormatterCallback,
  })  : assert(steps != null && steps.isNotEmpty,
            'There needs to be at least one step.'),
        assert(steps.every((it) => it >= min && it <= max),
            'Each step needs to be within slider values range.'),
        assert(initialStep == null || steps.contains(initialStep),
            'Initial step does not correspond to any of slider steps.'),
        super(key: key);

  // Key object passed to internal slider widget.
  final Key sliderKey;

  /// Slider values that the widget will snap to.
  final Set<double> steps;

  /// Slider value used when the widget's state is created for the first time.
  final double initialStep;

  /// Callback function invoked every time slider value gets snapped to adjacent step.
  final ValueChanged<double> onStepChanged;

  /// Whether the widget should display only step values or all the values
  /// between [min] and [max] (but won't snap to them regardless).
  final bool hardSnap;

  /// Modifies the way in which slider will snap to steps:
  /// * value - snaps to step if the value becomes less or equals to given value.
  /// * percent - snaps to step if the value becomes less or equals to given percent of [min]-to-[max] range.
  /// * stretch - snaps to step that is closest to slider's current value.
  final SnapMode snapMode;

  /// Slider's value animation curve.
  final Curve animCurve;

  /// Slider's value animation duration.
  final Duration animDuration;

  /// See [Slider.min].
  final double min;

  /// See [Slider.max].
  final double max;

  /// See [Slider.label].
  final String label;

  /// See [Slider.divisions].
  final int divisions;

  /// See [Slider.activeColor].
  final Color activeColor;

  /// See [Slider.inactiveColor].
  final Color inactiveColor;

  /// See [Slider.onChanged].
  final ValueChanged<double> onChanged;

  /// See [Slider.onChangeEnd].
  final ValueChanged<double> onChangeEnd;

  /// See [Slider.onChangeStart].
  final ValueChanged<double> onChangeStart;

  /// See [Slider.semanticFormatterCallback].
  final SemanticFormatterCallback semanticFormatterCallback;

  @override
  _StepSliderState createState() => _StepSliderState();
}

class _StepSliderState extends State<StepSlider>
    with SingleTickerProviderStateMixin {
  AnimationController _animator;
  CurvedAnimation _baseAnim;
  Animation<double> _animation;
  double _currentStep;
  Snapper _snapper;

  @override
  void didUpdateWidget(StepSlider oldWidget) {
    super.didUpdateWidget(oldWidget);
    final snapper = _createSnapper();
    if (snapper != _snapper) {
      _snapper = snapper;
    }
    _animator.duration = widget.animDuration;
    _baseAnim.curve = widget.animCurve;
  }

  @override
  void initState() {
    super.initState();
    _currentStep = widget.initialStep ?? widget.steps.first;
    _animator = AnimationController(
        vsync: this, duration: widget.animDuration, value: 1.0);
    _baseAnim = CurvedAnimation(parent: _animator, curve: widget.animCurve);
    _snapper = _createSnapper();
    _animateTo(_currentStep, false);
  }

  Snapper _createSnapper() => (widget.snapMode ?? SnapMode.stretch())
      .snapperFor(widget.steps, widget.min, widget.max);

  @override
  void dispose() {
    _animator.dispose();
    super.dispose();
  }

  @override
  Widget build(BuildContext context) {
    return AnimatedBuilder(
      animation: _animator,
      builder: (_, __) => Slider(
            key: widget.sliderKey,
            min: widget.min,
            max: widget.max,
            label: widget.label,
            divisions: widget.divisions,
            activeColor: widget.activeColor,
            inactiveColor: widget.inactiveColor,
            semanticFormatterCallback: widget.semanticFormatterCallback,
            value: _animation.value.clamp(widget.min, widget.max),
            onChangeStart: (it) {
              _onSliderChangeStart(it);
              widget.onChangeStart?.call(it);
            },
            onChangeEnd: (it) {
              _onSliderChangeEnd(it);
              widget.onChangeEnd?.call(it);
            },
            onChanged: (it) {
              _onSliderChanged(it);
              widget.onChanged?.call(it);
            },
          ),
    );
  }

  void _onSliderChangeStart(double value) {
    if (!widget.hardSnap) {
      _animateTo(value, true);
    }
  }

  void _onSliderChangeEnd(double value) {
    if (!widget.hardSnap) {
      _animateTo(_currentStep, true);
    }
  }

  void _onSliderChanged(double value) {
    if (!widget.hardSnap) {
      _animateTo(value, false);
    }
    final step = _snapper.snap(value);
    if (step != null && step != _currentStep) {
      _onStepChanged(step);
    }
  }

  void _onStepChanged(double step) {
    _currentStep = step;
    if (widget.hardSnap) {
      _animateTo(_currentStep, true);
    }
    widget.onStepChanged?.call(_currentStep);
  }

  void _animateTo(double end, bool restart) {
    _animation =
        Tween(begin: _animation?.value ?? end, end: end).animate(_baseAnim);
    if (restart) {
      _animator.forward(from: 0.0);
    }
    if (!_animator.isAnimating) {
      setState(() {});
    }
  }
}


lib/src/snapper.dart
import 'dart:math';

import 'package:collection/collection.dart';
import 'package:quiver_hashcode/hashcode.dart';

abstract class Snapper {
  SnapperConfig _config;
  set config(SnapperConfig value) {
    _config = value;
    _boundsMap = null;
  }

  List<double> _getBoundsList();

  double snap(double value) => (_boundsMap ??= _getBoundsMap())
      .entries
      .firstWhere((it) => it.value.contains(value), orElse: () => null)
      ?.key;

  Map<double, Range> _boundsMap;
  Map<double, Range> _getBoundsMap() {
    final boundsMap = <double, Range>{};
    final boundsList = _getBoundsList();
    var index = 0;
    _config.items.forEach((it) {
      boundsMap[it] = Range(boundsList[2 * index], boundsList[2 * index + 1]);
      ++index;
    });
    return boundsMap;
  }

  List<double> _getBoundsForDelta(double delta) {
    final bounds = <double>[];
    bounds.add(max(_config.min, _config.items.first - delta));
    _config.items.reduce((a, b) {
      var rangeA = a + delta;
      var rangeB = b - delta;
      if (rangeA > rangeB) {
        rangeA = (rangeA + rangeB) / 2;
        rangeB = rangeA;
      }
      bounds..add(rangeA)..add(rangeB);
      return b;
    });
    bounds.add(min(_config.max, _config.items.last + delta));
    return bounds;
  }

  @override
  bool operator ==(o) => o != null && o is Snapper && o._config == this._config;

  @override
  int get hashCode => _config.hashCode;
}

class ValueSnapper extends Snapper {
  ValueSnapper(this.value);

  final double value;

  List<double> _getBoundsList() => _getBoundsForDelta(value);

  @override
  bool operator ==(o) =>
      o != null && o is ValueSnapper && o.value == this.value;

  @override
  int get hashCode => value.hashCode;
}

class PercentSnapper extends Snapper {
  PercentSnapper(this.percent);

  final double percent;

  @override
  List<double> _getBoundsList() =>
      _getBoundsForDelta((_config.max - _config.min) * percent);

  @override
  bool operator ==(o) =>
      o != null && o is PercentSnapper && o.percent == this.percent;

  @override
  int get hashCode => percent.hashCode;
}

class StretchSnapper extends Snapper {
  List<double> _getBoundsList() {
    final bounds = <double>[];
    bounds.add(_config.min);
    _config.items.reduce((a, b) {
      final range = (a + b) / 2;
      bounds..add(range)..add(range);
      return b;
    });
    bounds.add(_config.max);
    return bounds;
  }

  @override
  bool operator ==(o) => o != null && o is StretchSnapper;

  @override
  int get hashCode => super.hashCode;
}

class SnapperConfig {
  SnapperConfig(this.items, this.min, this.max);
  final Set<double> items;
  final double min;
  final double max;

  @override
  bool operator ==(o) =>
      o != null &&
      o is SnapperConfig &&
      o.min == this.min &&
      o.max == this.max &&
      SetEquality<double>().equals(o.items, this.items);

  @override
  int get hashCode => hash3(items, min, max);
}

class Range {
  Range(this.begin, this.end);
  final double begin;
  final double end;
  bool contains(double value) => value >= begin && value <= end;
}


example/main.dart
import 'package:flutter/material.dart';
import 'package:step_slider/slider.dart';

void main() => runApp(SliderExample());

class SliderExample extends StatefulWidget {
  @override
  _SliderExampleState createState() => _SliderExampleState();
}

class _SliderExampleState extends State<SliderExample> {
  double _currentStep = 150.0;
  SnapMode _snapMode = SnapMode.stretch();
  bool _hardSnap = false;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'StepSlider example',
      home: Scaffold(
        appBar: AppBar(title: Text('StepSlider example')),
        body: Padding(
          padding: EdgeInsets.all(16.0),
          child: Column(children: <Widget>[
            StepSlider(
              min: 100.0,
              max: 200.0,
              steps: {110, 150, 160, 195},
              initialStep: _currentStep,
              animCurve: Curves.ease,
              animDuration: Duration(seconds: 1),
              snapMode: _snapMode,
              hardSnap: _hardSnap,
              onStepChanged: (it) {
                setState(() => _currentStep = it);
              },
              // ... slider's other args
            ),
            Center(child: Text('Current step: $_currentStep')),
            Container(height: 32.0),
            Row(mainAxisSize: MainAxisSize.min, children: <Widget>[
              Padding(
                padding: EdgeInsets.symmetric(horizontal: 32.0),
                child: Center(child: Text('Hard snap')),
              ),
              Switch(
                value: _hardSnap,
                onChanged: (it) {
                  setState(() => _hardSnap = it);
                },
              )
            ]),
            Container(height: 16.0),
            Row(mainAxisSize: MainAxisSize.min, children: <Widget>[
              Padding(
                padding: EdgeInsets.symmetric(horizontal: 32.0),
                child: Center(child: Text('Snap mode')),
              ),
              DropdownButton<SnapMode>(
                value: _snapMode,
                onChanged: (it) {
                  setState(() => _snapMode = it);
                },
                items: [
                  DropdownMenuItem(
                    value: SnapMode.value(5),
                    child: Text('Value'),
                  ),
                  DropdownMenuItem(
                    value: SnapMode.percent(10),
                    child: Text('Percent'),
                  ),
                  DropdownMenuItem(
                    value: SnapMode.stretch(),
                    child: Text('Stretch'),
                  ),
                ],
              )
            ]),
          ]),
        ),
      ),
    );
  }
}
